File: src/data/courses/dsa/trie/trie-design-variations/dictionary-wildcard/code/python/dictionary_wildcard.py
# ============================================================
# Solution 1: Brute Force — Linear scan all words O(N*L)
# ============================================================
class Solution1:
    def __init__(self): self.words = []
    def addWord(self, word): self.words.append(word)
    def search(self, query):
        result = []
        for w in self.words:
            if len(w) != len(query): continue
            if all(q == '.' or q == c for q, c in zip(query, w)):
                result.append(w)
        return result

# ============================================================
# Solution 1: Brute Force
# ============================================================

class Solution1:
    # Brute force: hash set / nested loops / direct comparison
    # See Solution 2 below for optimal Trie-based approach
    pass


# ============================================================
# Solution 2: Optimal (Trie-based)
# ============================================================
class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end = False

class WordDictionary:
    def __init__(self):
        self.root = TrieNode()
    
    def addWord(self, word):
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.is_end = True
    
    def search(self, word):
        return self._search_helper(word, 0, self.root)
    
    def _search_helper(self, word, index, node):
        if index == len(word):
            return node.is_end
        
        char = word[index]
        if char == '.':
            for child in node.children.values():
                if self._search_helper(word, index + 1, child):
                    return True
            return False
        else:
            if char not in node.children:
                return False
            return self._search_helper(word, index + 1, node.children[char])

File: src/data/courses/dsa/trie/trie-design-variations/dictionary-wildcard/code/cpp/DictionaryWildcard.cpp
// ============================================================
// Solution 1: Store words in vector, linear scan with matching
// Time: O(N * L) per search | Space: O(N * L)
// ============================================================
#include <string>
#include <vector>
using namespace std;

class Solution1 {
    vector<string> words;
public:
    void addWord(string word) { words.push_back(word); }
    bool search(string pattern) {
        for (auto& w : words) {
            if (w.size() != pattern.size()) continue;
            bool ok = true;
            for (int i = 0; i < (int)w.size(); i++)
                if (pattern[i] != '.' && pattern[i] != w[i]) { ok = false; break; }
            if (ok) return true;
        }
        return false;
    }
};

// ============================================================
// Solution 2: Optimal (Trie-based)
// ============================================================
#include <string>
#include <unordered_map>
using namespace std;

// Time: O(L) add, O(26^W*L) search | Space: O(N*L)

struct TrieNode {
    unordered_map<char, TrieNode*> children;
    bool isEnd = false;
};

class Solution2 {
    TrieNode* root;
    
    bool searchHelper(const string& word, int index, TrieNode* node) {
        if (index == word.size()) return node->isEnd;
        
        char ch = word[index];
        if (ch == '.') {
            for (auto& [c, child] : node->children) {
                if (searchHelper(word, index + 1, child)) return true;
            }
            return false;
        } else {
            if (!node->children[ch]) return false;
            return searchHelper(word, index + 1, node->children[ch]);
        }
    }
    
public:
    WordDictionary() {
        root = new TrieNode();
    }
    
    void addWord(string word) {
        TrieNode* node = root;
        for (char ch : word) {
            if (!node->children[ch]) {
                node->children[ch] = new TrieNode();
            }
            node = node->children[ch];
        }
        node->isEnd = true;
    }
    
    bool search(string word) {
        return searchHelper(word, 0, root);
    }
};
